### DISCUSSION TITLE
IsLastItemConverter / IsNotLastItemConverter

### DISCUSSION THREAD

**Discussion by ewerspej** [CommunityToolkit/Maui#1339] [New Feature Discussions]
I would like to propose to add an `IsLastItemConverter` as well as an `IsNotLastItemConverter` to the list of converters to determine whether or not an element is the last item in an `IEnumerable` and then update the view accordingly.

The implementation could be as simple as:

```c#
public class IsLastItemConverter : IValueConverter
{
    public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
    {
        return parameter is IEnumerable<object> enumerable && enumerable.Last().Equals(value);
    }

    public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
    {
        return Binding.DoNothing;
    }
}
```

```c#
public class IsNotLastItemConverter : IValueConverter
{
    public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
    {
        return parameter is IEnumerable<object> enumerable && !enumerable.Last().Equals(value);
    }

    public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
    {
        return Binding.DoNothing;
    }
}
```

One major use case would be to allow users to make decisions in their XAML based on whether an item is the last element in an `ItemsSource`, for example to show a divider or use a different color value, etc:

```xml
<ContentPage.Resources>
    <ResourceDictionary>
        <mct:IsNotLastItemConverter x:Key="IsNotLastItemConverter" />
    </ResourceDictionary>
</ContentPage.Resources>

<VerticalStackLayout 
    BindableLayout.ItemsSource="{Binding Items}">
    <BindableLayout.ItemTemplate>
        <DataTemplate x:DataType="model:MyClass">
            <Grid RowDefinitions="40,auto">
                <Label
                    Grid.Row="0"
                    Text="{Binding Name}" />
                <BoxView
                    Grid.Row="1"
                    HeightRequest="1"
                    IsVisible="{Binding ., Converter={StaticResource IsNotLastItemConverter}, ConverterParameter={Binding Items, Source={RelativeSource AncestorType={x:Type viewModel:MyViewModel}}}}" />
            </Grid>
        <DataTemplate>
    </BindableLayout.ItemTemplate>
</VerticalStackLayout>
```

I can provide a PR with this myself, if this is something that is deemed useful.

**Notes:**
1. I am aware that this will only work with reference types, but that's because the `IValueConverter` interface doesn't allow to specify a generic type argument. Alternatively, I would need to use `IEnumerable` without the type and iterate through the enumerable until I hit the last item and compare that to the `value` argument.
2. One problem could be that items can appear multiple times in lists and collections which could lead to unexpected behavior

**Comment by ewerspej** [comment]
I don't think this is possible, because `ConverterParameter` is just a property of type `object` and not a `BindableProperty`.

**Reply by pictos** [reply]
So, the Converter is inside the Binding (look you snippet), with that everytime that property changes the new value will be send to the Converter to perform the check

**Reply by ewerspej** [reply]
it still won't work, because the parameter is not passed as the bound value, it's passed as a `Binding` instance here:

```c#
ConverterParameter={Binding Items, Source={RelativeSource AncestorType={x:Type viewModel:MyViewModel}}}
```

**Reply by ewerspej** [reply]
@pictos The intended behavior can be achieved using a `MultiValueConverter`, which I've just done in a project of mine.
